packages/typescript-estree/src/createParserServices.ts

```diff
@@ -0,0 +1,27 @@
+import type * as ts from 'typescript';
+
+import type { ASTMaps } from './convert';
+import type { ParserServices } from './parser-options';
+
+export function createParserServices(
+  astMaps: ASTMaps,
+  program: ts.Program | null,
+): ParserServices {
+  if (!program) {
+    // we always return the node maps because
+    // (a) they don't require type info and
+    // (b) they can be useful when using some of TS's internal non-type-aware AST utils
+    return { program, ...astMaps };
+  }
+
+  const checker = program.getTypeChecker();
+
+  return {
+    program,
+    ...astMaps,
+    getSymbolAtLocation: node =>
+      checker.getSymbolAtLocation(astMaps.esTreeNodeToTSNodeMap.get(node)),
+    getTypeAtLocation: node =>
+      checker.getTypeAtLocation(astMaps.esTreeNodeToTSNodeMap.get(node)),
```

Comment:
do we want to do a weakmap memo on these?
Eg
```
function memoize<TNode, TRet>(fn: (node: TNode) => TRet): (node: TNode) => TRet {
  const cache = new WeakMap<TNode, TRet>();

  return (node) => {
    const cachedRet = cache.get(node);
    if (cachedRet) {
      return cachedRet;
    }

    const newRet = fn(node);
    cachedRet = cache.set(node, newRet);
    return newRet;
  };
}
```

That would allow us to enforce that if two rules request the exact same node then we'll 100% use a fast path.

Doing a quick look at the implementation of `getTypeAtLocation` - it doesn't do any fast caching, so it does do a bunch of work.

IDK if this sort of thing would actually save any real time though - would be worth a test. Maybe we can file an issue to look at this later?

Reply:
Yeah I'll file an issue and do some perf measuring. Exciting!

Reply:
#6425